### PR TITLE
1115/healthcheck

### LINKED ISSUE #1115: Health check API endpoint
How about an unauthenticated API endpoint that will return the health status of various components on which a successful authentication is dependent?
For example:

GET /system/health

Token database : OK
Audit database: OK
LDAP resolver: OK
HSM connection: OK

### PR DESCRIPTION
**Pull request by henrik-falk**

### CODE REVIEW

File: privacyidea/api/healthcheck.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+healthz_blueprint = Blueprint('healthz_blueprint', __name__)
+
+
+@healthz_blueprint.route('/', methods=['GET'])
+def healthz():
+    """
+    Health check endpoint that indicates if the app is healthy (running and ready to serve requests).
+    """
+    app = current_app._get_current_object()
+    if not app.config.get('APP_READY'):
```
Comment (plettich) [inline privacyidea/api/healthcheck.py:19]: so currently, when the endpoint is called, the app must already exist. But then the `APP_READY` is always true (unless it set to false again somewhere else).
I guess it is a chicken-egg problem...

File: privacyidea/api/healthcheck.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        return send_result({"status": "not healthy"}), 503
+    elif not get_hsm().is_ready:
+        return send_result({"status": "not healthy"}), 503
```
Comment (plettich) [inline privacyidea/api/healthcheck.py:22]: can we return some additional information here, that the app is not healthy due to the HSM? Or is that not intended?

**Review by plettich — CHANGES_REQUESTED** [review]

File: privacyidea/api/healthcheck.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
+from flask import Blueprint, current_app
```
[the remark this answers is not in the document]
Comment (henrik-falk) [reply privacyidea/api/healthcheck.py:32]: fixed with commit 680e7655f0874d02c883a1fdcb4dc6647c3104df

[the remark this answers is not in the document]
Comment (henrik-falk) [reply privacyidea/api/healthcheck.py:32]: fixed with commit 680e7655f0874d02c883a1fdcb4dc6647c3104df

File: privacyidea/api/healthcheck.py [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+healthz_blueprint = Blueprint('healthz_blueprint', __name__)
+
+
+@healthz_blueprint.route('/', methods=['GET'])
+def healthz():
+    """
+    Health check endpoint that indicates if the app is healthy (running and ready to serve requests).
+    """
+    app = current_app._get_current_object()
```
[the remark this answers is not in the document]
Comment (henrik-falk) [reply privacyidea/api/healthcheck.py]: fixed with commit 680e7655f0874d02c883a1fdcb4dc6647c3104df

File: privacyidea/api/healthcheck.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+        return send_result(result), 200
+    except Exception as e:
+        log.debug(f"Exception in /resolversz endpoint: {e}")
+        return send_result({"status": "error"}), 500
+
+
+def check_resolvers(resolvers_list):
+    """
+    Test the connections for the given resolver list.
+    Args:
```
[the remark this answers is not in the document]
Comment (henrik-falk) [reply privacyidea/api/healthcheck.py]: fixed with commit 680e7655f0874d02c883a1fdcb4dc6647c3104df

File: privacyidea/api/healthcheck.py [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+        ldapresolvers_list = get_resolver_list(filter_resolver_type="ldapresolver")
+        sqlresolvers_list = get_resolver_list(filter_resolver_type="sqlresolver")
+
+        result["ldapresolvers"] = check_resolvers(ldapresolvers_list)
+        result["sqlresolvers"] = check_resolvers(sqlresolvers_list)
+        result["status"] = "OK"
+        return send_result(result), 200
+    except Exception as e:
+        log.debug(f"Exception in /resolversz endpoint: {e}")
+        return send_result({"status": "error"}), 500
```
[the remark this answers is not in the document]
Comment (henrik-falk) [reply privacyidea/api/healthcheck.py]: fixed with commit 680e7655f0874d02c883a1fdcb4dc6647c3104df

File: privacyidea/api/healthcheck.py (hunk 1 above)
Comment (henrik-falk) [reply privacyidea/api/healthcheck.py]: Correct, the APP_READY flag is currently a placeholder until a more sophisticated method of checking if the server has started is developed. In the future, this flag should be switched to False every time dependencies anywhere in the code are not ready.

File: privacyidea/api/healthcheck.py (hunk 2 above)
Comment (henrik-falk) [reply privacyidea/api/healthcheck.py]: fixed with commit 680e7655f0874d02c883a1fdcb4dc6647c3104df

**Review by nilsbehlen — APPROVED** [review]

**Pull request merged**